lib/buffer.js

```diff
@@ -235,27 +238,45 @@ function copyImpl(source, target, targetStart, sourceStart, sourceEnd) {
       throw new ERR_OUT_OF_RANGE('sourceEnd', '>= 0', sourceEnd);
   }
 
+  // Clamp sourceEnd to be inbounds of source buffer.
+  // This is expected behavior and not to throw.
+  sourceEnd = MathMin(sourceEnd, source.byteLength);
+
   if (targetStart >= target.byteLength || sourceStart >= sourceEnd)
     return 0;
 
-  return _copyActual(source, target, targetStart, sourceStart, sourceEnd);
-}
-
-function _copyActual(source, target, targetStart, sourceStart, sourceEnd) {
-  if (sourceEnd - sourceStart > target.byteLength - targetStart)
-    sourceEnd = sourceStart + target.byteLength - targetStart;
+  // Assert source slice in bounds of source buffer.
+  // 0 <= sourceStart < sourceEnd <= source.byteLength
+  assert(0 <= sourceStart && sourceStart < sourceEnd && sourceEnd <= source.byteLength);
 
-  let nb = sourceEnd - sourceStart;
-  const sourceLen = source.byteLength - sourceStart;
-  if (nb > sourceLen)
-    nb = sourceLen;
+  // Assert target slice in bounds of target buffer.
+  // 0 <= targetStart < target.byteLength
+  assert(0 <= targetStart && targetStart < target.byteLength);
 
-  if (nb <= 0)
-    return 0;
+  // Truncate source so that its length doesn't exceed targets.
+  // This is the expected behavior, not to throw.
+  const copyLength = MathMin(sourceEnd - sourceStart, target.byteLength - targetStart);
+  sourceEnd = sourceStart + copyLength;
 
-  _copy(source, target, targetStart, sourceStart, nb);
+  return _copyActual(source, target, targetStart, sourceStart, sourceStart + copyLength);
+}
 
-  return nb;
+// Safely performs native copy from valid source slice to valid target slice.
+// - The source slice is not clamped to fit into the target slice. If it won't fit, this throws.
+// - If either the source or target slice are out of bounds, this throws.
+function _copyActual(source, target, targetStart, sourceStart, sourceEnd) {
+  assert(isUint8Array(source) && isUint8Array(target));
```

Comment:
The assert here is a bit redundant given the checks that are made in the `copyImpl` call site.

Reply:
It is technically redundant, but asserts usually are.

In my opinion this assert is useful since it makes `_copyActual`'s invariant that `source` and `target` need to be `Uint8Array` explicit. 

It indicates to future/new maintainer that will want to call `_copyActual`, what the invariants are. They don't have to guess at the invariants, or deduce them by exploring all the call-sites of `_copyActual`.